File: face/geomajas-face-dojo/client/src/main/resources/geomajas/spatial/geometry/MultiLineString.js
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

dojo.provide("geomajas.spatial.geometry.MultiLineString");
dojo.declare("MultiLineString", Geometry, {

	/**
	 * @fileoverview Geometry implementation for MultiLineStrings.
	 * @class Geometry implementation for MultiLineStrings.
	 * @author Pieter De Graef
	 *
	 * @constructor
	 * @extends Geometry
	 */
	constructor : function () {
		this.lineStrings = [];
	},

	/**
	 * Create a copy of this geometry and return it.
	 */
	clone : function () {
		var fac = this.getGeometryFactory();
		var lines = [];
		for(var i=0; i<this.lineStrings.length; i++) {
			lines[i] = this.lineStrings[i].clone();
		}
		return fac.createMultiLineString(lines);
	},

	/**
	 * Multiple cases. If n is a integer, the n'th coordinate in the exterior
	 * ring will be returned. If n is an array with 2 indeces, the first integer
	 * depicts the linearring, the second integer the index in this ring.
	 * @param n Should be an array with 2 integers. Although just an integer is 
	 *          also supported.
	 */
	getCoordinateN : function (n) {
		if (n instanceof Array && n.length == 1) { // assume [0, n[0]]
			if (this.lineStrings.length >= 1) {
				var lineString = this.lineStrings[0];
				return lineString.getCoordinateN(n);
			}
		} else if (n instanceof Array && n.length >= 2) { // assume [n[0], n[1]]
			if (this.lineStrings.length >= 1) {
				var lineString = this.lineStrings[n[0]];
				return lineString.getCoordinateN(n[1]);
			}
		} else {
			if (this.lineStrings.length >= 1 && n != null) {
				var lineString = this.lineStrings[0];
				return lineString.getCoordinateN(n);
			}
		}
		return this;
	},

	/**
	 * Multiple cases. If n is an integer, the n'th {@link LineString} will
	 * be returned. If n  is an array of length 1 same as above. If n is an
	 * array of length 2, this function will delegate to the linestring n[0].
	 * It will ask for the "getGeometryN(n[1])".
	 * @param n Should be an array with max 2 integers. Although just an
	 *          integer is also supported.
	 */
	getGeometryN : function (n) {
		if (n instanceof Array && n.length >= 2) {
			return this.lineStrings[n[0]].getGeometryN(n[1]);
		} else if (n instanceof Array && n.length == 1) {
			return this.lineStrings[n[0]];
		} else if (n >= 0 && n < this.lineStrings.length) {
			return this.lineStrings[n];
		}
		
		return null;
	},

	/**
	 * Returns the number of linestrings.
	 */
	getNumGeometries : function () {
		return this.lineStrings.length;
	},

	/**
	 * Returns 0.
	 */
	getArea : function () {
		return 0;
	},

	/**
	 * Returns the added length of all the linestrings.
	 */
	getLength : function () {
		var len = 0;
		for (var i=0; i<this.lineStrings.length; i++) {
			len += this.lineStrings[i].getLength();
		}
		return len;
	},

	/**
	 * The centroid is also known as the "centre of gravity" or the
	 * "center of mass".
	 * @return Return the centre point.
	 */
	getCentroid : function () {
		return this.lineStrings[0].getCentroid();
	},

	/**
	 * Returns the minimal distance between this coordinate and any vertex of the geometry
	 * @return Return the minimal distance
	 */
	getDistance : function (coordinate) {
		var distance = Number.MAX_VALUE;
		for (var i=0; i< this.lineStrings.length; i++) {
			var d = this.lineStrings[i].getDistance(coordinate);
			if(d < distance){
				distance = d;
			}
		}
		return distance;		
	},

	getDistanceApprox : function (coordinate) {
		var distance = Number.MAX_VALUE;
		for (var i=0; i< this.lineStrings.length; i++) {
			var d = this.lineStrings[i].getDistanceApprox(coordinate);
			if(d < distance){
				distance = d;
			}
		}
		return distance;		
	},

	/**
	 * Check the validity of all the linestrings.
	 */
	isValid : function () {
		for (var i=0; i<this.lineStrings.length; i++) {
			if (!this.lineStrings[i].isValid()) {
				return false;
			}
		}
		return true;
	},

	/**
	 * Shouldn't be here!
	 */
	appendCoordinate : function (coordinate) {
		if (this.lineStrings == null || this.lineStrings.length == 0) {
			var fac = new GeometryFactory(this.srid, this.precision);
			this.lineStrings = [ fac.createLineString([coordinate]) ];
		} else {
			this.lineStrings[0].appendCoordinate(coordinate);
		}
	},

	/**
	 * Returns the closest Bbox around the geometry.
	 */
	getBounds : function () {
		var num = this.getNumGeometries();
		if (num == null || num == 0) {
			return null;
		}
		var bounds = null;
		for (var i=0; i<num; i++) {
			var coords = this.getGeometryN(i).getCoordinates();
			if (bounds == null) {
				bounds = this._computeBounds(coords);
			} else {
				bounds = bounds.union(this._computeBounds(coords));
			}
		}
		return bounds;
	},

	/**
	 * @private
	 */
	_getPrimaryCoordinates : function () {
		if (this.lineStrings != null && this.lineStrings.length > 0) {
			return this.lineStrings[0].getCoordinates();
		}
		return null;
	}
});

File: face/geomajas-gwt-face/client/src/main/java/org/geomajas/gwt/client/spatial/geometry/MultiLineString.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.gwt.client.spatial.geometry;

import org.geomajas.geometry.Coordinate;
import org.geomajas.gwt.client.spatial.Bbox;
import org.geomajas.layer.LayerType;

/**
 * MultiLineString client-side GWT object.
 *
 * @author Pieter De Graef
 */
public class MultiLineString extends AbstractGeometry {

	private static final long serialVersionUID = 5319457800593690343L;

	private LineString[] lineStrings;

	// -------------------------------------------------------------------------
	// Constructor (package visibility)
	// -------------------------------------------------------------------------

	/**
	 * Constructor using srid and precision.
	 *
	 * @param srid srid
	 * @param precision precision
	 */
	MultiLineString(int srid, int precision) {
		super(srid, precision);
	}

	/**
	 * Constructor using srid, precision and the linestrings to combine.
	 *
	 * @param srid srid
	 * @param precision precision
	 * @param lineStrings linestrings to combine
	 */
	MultiLineString(int srid, int precision, LineString[] lineStrings) {
		super(srid, precision);
		this.lineStrings = lineStrings;
	}

	// -------------------------------------------------------------------------
	// Geometry implementation:
	// -------------------------------------------------------------------------

	@Override
	public Object clone() { // NOSONAR super.clone() not supported by GWT
		return getGeometryFactory().createMultiLineString(lineStrings);
	}

	@Override
	public Geometry getGeometryN(int n) {
		if (!isEmpty()) {
			if (n >= 0 && n < lineStrings.length) {
				return lineStrings[n];
			}
		}
		return this;
	}

	@Override
	public int getNumGeometries() {
		if (isEmpty()) {
			return 0;
		}
		return lineStrings.length;
	}

	@Override
	public int getNumPoints() {
		if (isEmpty()) {
			return 0;
		}
		int sum = 0;
		for (LineString lineString : lineStrings) {
			sum += lineString.getNumPoints();
		}
		return sum;
	}

	LineString[] getLineStrings() {
		return lineStrings;
	}

	void setLineStrings(LineString[] lineStrings) {
		this.lineStrings = lineStrings;
	}

	@Override
	public double getLength() {
		double len = 0;
		if (!isEmpty()) {
			for (LineString lineString : lineStrings) {
				len += lineString.getLength();
			}
		}
		return len;
	}

	@Override
	public Coordinate getCentroid() {
		if (isEmpty()) {
			return null;
		}
		double sumX = 0;
		double sumY = 0;
		double totalLength = 0;
		Coordinate[] coordinates = new Coordinate[lineStrings.length];
		for (int i = 0; i < lineStrings.length; i++) {
			coordinates[i] = lineStrings[i].getCentroid();
		}
		for (int i = 0; i < coordinates.length - 1; i++) {
			double length = coordinates[i].distance(coordinates[i + 1]);
			totalLength += length;
			double midx = (coordinates[i].getX() + coordinates[i + 1].getX()) / 2;
			sumX += length * midx;
			double midy = (coordinates[i].getY() + coordinates[i + 1].getY()) / 2;
			sumY += length * midy;
		}
		return new Coordinate(sumX / totalLength, sumY / totalLength);
	}

	/**
	 * Return the minimal distance between this coordinate and any vertex of the geometry.
	 *
	 * @return Return the minimal distance
	 */
	public double getDistance(Coordinate coordinate) {
		double distance = Double.MAX_VALUE;
		if (!isEmpty()) {
			for (LineString lineString : lineStrings) {
				double d = lineString.getDistance(coordinate);
				if (d < distance) {
					distance = d;
				}
			}
		}
		return distance;
	}

	@Override
	public boolean isEmpty() {
		return (lineStrings == null || lineStrings.length == 0);
	}

	@Override
	public boolean isValid() {
		if (isEmpty()) {
			return true;
		}
		for (LineString lineString : lineStrings) {
			if (!lineString.isValid()) {
				return false;
			}
		}
		return true;
	}

	@Override
	public boolean isSimple() {
		if (isEmpty()) {
			return true;
		}
		for (int i = 0; i < lineStrings.length - 1; i++) {
			for (int j = i + 1; j < lineStrings.length; j++) {
				if (lineStrings[i].intersects(lineStrings[j])) {
					return false;
				}
			}
		}

		return true;
	}

	@Override
	public Bbox getBounds() {
		Bbox bounds = null;
		if (!isEmpty()) {
			for (LineString lineString : lineStrings) {
				if (bounds == null) {
					bounds = lineString.getBounds();
				} else {
					bounds = bounds.union(lineString.getBounds());
				}
			}
		}
		return bounds;
	}

	@Override
	public Coordinate getCoordinate() {
		if (isEmpty()) {
			return null;
		}
		return lineStrings[0].getCoordinate();
	}

	/**
	 * Get the full concatenated list of coordinates of all the LineStrings in this MultiLineString geometry.
	 */
	public Coordinate[] getCoordinates() {
		if (isEmpty()) {
			return null;
		}
		Coordinate[] coordinates = new Coordinate[getNumPoints()];
		int count = 0;
		for (LineString lineString : lineStrings) {
			for (int j = 0; j < lineString.getNumPoints(); j++) {
				coordinates[count++] = lineString.getCoordinateN(j);
			}
		}
		return coordinates;
	}

	@Override
	public LayerType getLayerType() {
		return LayerType.MULTILINESTRING;
	}

	/**
	 * Check the LineStrings one by one for intersections with the given geometry.
	 */
	public boolean intersects(Geometry geometry) {
		if (isEmpty()) {
			return false;
		}
		for (LineString lineString : lineStrings) {
			if (lineString.intersects(geometry)) {
				return true;
			}
		}

		return false;
	}

	@Override
	public String toWkt() {
		if (isEmpty()) {
			return "MULTILINESTRING EMPTY";
		}
		String wkt = "MULTILINESTRING (";
		for (int i = 0; i < lineStrings.length; i++) {
			String lineWkt = lineStrings[i].toWkt();
			if (i > 0) {
				wkt += ",";
			}
			wkt += lineWkt.substring(lineWkt.indexOf("("));
		}
		return wkt + ")";
	}
}
